src/main/java/org/broadinstitute/consent/http/db/SamDAO.java

```diff
@@ -54,10 +54,17 @@ public SamDAO(HttpClientUtil clientUtil, ServicesConfiguration configuration) {
   public List<ResourceType> getResourceTypes(AuthUser authUser) throws Exception {
     GenericUrl genericUrl = new GenericUrl(configuration.getV1ResourceTypesUrl());
     HttpRequest request = clientUtil.buildGetRequest(genericUrl, authUser);
-    HttpResponse response = executeRequest(request);
-    if (!response.isSuccessStatusCode()) {
-      logException("Error getting resource types from Sam: " + response.getStatusMessage(),
-          new ServerErrorException(response.getStatusMessage(), response.getStatusCode()));
+    HttpResponse response;
+    try {
+      response = executeRequest(request);
+      if (!response.isSuccessStatusCode()) {
+        logException("Error getting resource types from Sam: " + response.getStatusMessage(),
+            new ServerErrorException(response.getStatusMessage(), response.getStatusCode()));
+      }
```

Comment:
I believe this can be done using a generic wrapper API to avoid the code duplication. All the error-wrapped APIs are of type `Function<T, R>` except `getToSText()` which is `Supplier<T>`. Something like
```
  interface FunctionWithError<T, R> {
    R apply(T t) throws ServerErrorException;
  }

  private <T, R> R handleError(FunctionWithError<T, R> wrapped, Function<T, R> onError, T input,
      Supplier<String> logMessage) {
    try {
      return wrapped.apply(input);
    } catch (ServerErrorException e) {
      logException("Sam is down, returning %s".formatted(logMessage.get()), e);
      return onError.apply(input);
    }
  }
```

Reply:
Thanks for the suggestion, I'll dig into that